File: elixir/lib/algorithms/dfs.ex
defmodule Algorithms.DFS do
  @moduledoc """
  Depth-first search algorithms.
  Time: O(V + E)
  Space: O(V)
  """

  alias Algorithms.Graph

  @doc "DFS traversal from a starting vertex (iterative)"
  @spec traverse(Graph.t(), any()) :: [any()]
  def traverse(graph, start) do
    if Graph.has_vertex?(graph, start) do
      do_traverse(graph, [start], MapSet.new(), [])
    else
      []
    end
  end

  defp do_traverse(_graph, [], _visited, result), do: Enum.reverse(result)

  defp do_traverse(graph, [vertex | stack], visited, result) do
    if MapSet.member?(visited, vertex) do
      do_traverse(graph, stack, visited, result)
    else
      new_visited = MapSet.put(visited, vertex)

      neighbors =
        graph
        |> Graph.neighbors(vertex)
        |> MapSet.to_list()
        |> Enum.filter(&(not MapSet.member?(new_visited, &1)))

      do_traverse(graph, neighbors ++ stack, new_visited, [vertex | result])
    end
  end

  @doc "DFS traversal from a starting vertex (recursive)"
  @spec traverse_recursive(Graph.t(), any()) :: [any()]
  def traverse_recursive(graph, start) do
    if Graph.has_vertex?(graph, start) do
      {result, _} = do_traverse_recursive(graph, start, MapSet.new())
      result
    else
      []
    end
  end

  defp do_traverse_recursive(graph, vertex, visited) do
    new_visited = MapSet.put(visited, vertex)

    {children_result, final_visited} =
      graph
      |> Graph.neighbors(vertex)
      |> MapSet.to_list()
      |> Enum.reduce({[], new_visited}, fn neighbor, {acc, vis} ->
        if MapSet.member?(vis, neighbor) do
          {acc, vis}
        else
          {result, new_vis} = do_traverse_recursive(graph, neighbor, vis)
          {acc ++ result, new_vis}
        end
      end)

    {[vertex | children_result], final_visited}
  end

  @doc "Find a path between two vertices using DFS"
  @spec find_path(Graph.t(), any(), any()) :: {:ok, [any()]} | {:error, :not_found}
  def find_path(graph, start, target) do
    cond do
      not Graph.has_vertex?(graph, start) -> {:error, :not_found}
      not Graph.has_vertex?(graph, target) -> {:error, :not_found}
      start == target -> {:ok, [start]}
      true -> do_find_path(graph, [{start, [start]}], MapSet.new(), target)
    end
  end

  defp do_find_path(_graph, [], _visited, _target), do: {:error, :not_found}

  defp do_find_path(graph, [{vertex, path} | stack], visited, target) do
    cond do
      vertex == target ->
        {:ok, Enum.reverse(path)}

      MapSet.member?(visited, vertex) ->
        do_find_path(graph, stack, visited, target)

      true ->
        new_visited = MapSet.put(visited, vertex)

        neighbors =
          graph
          |> Graph.neighbors(vertex)
          |> MapSet.to_list()
          |> Enum.filter(&(not MapSet.member?(new_visited, &1)))
          |> Enum.map(&{&1, [&1 | path]})

        do_find_path(graph, neighbors ++ stack, new_visited, target)
    end
  end

  @doc "Check if graph has a cycle (for directed graphs)"
  @spec has_cycle?(Graph.t()) :: boolean()
  def has_cycle?(graph) do
    vertices = Graph.vertices(graph)

    Enum.any?(vertices, fn vertex ->
      has_cycle_from?(graph, vertex, MapSet.new(), MapSet.new())
    end)
  end

  defp has_cycle_from?(graph, vertex, visited, rec_stack) do
    cond do
      MapSet.member?(rec_stack, vertex) ->
        true

      MapSet.member?(visited, vertex) ->
        false

      true ->
        new_visited = MapSet.put(visited, vertex)
        new_rec_stack = MapSet.put(rec_stack, vertex)

        result =
          graph
          |> Graph.neighbors(vertex)
          |> MapSet.to_list()
          |> Enum.any?(&has_cycle_from?(graph, &1, new_visited, new_rec_stack))

        result
    end
  end

  @doc "Topological sort (only for directed acyclic graphs)"
  @spec topological_sort(Graph.t()) :: {:ok, [any()]} | {:error, :has_cycle | :not_directed}
  def topological_sort(graph) do
    cond do
      not Graph.directed?(graph) -> {:error, :not_directed}
      has_cycle?(graph) -> {:error, :has_cycle}
      true -> {:ok, do_topological_sort(graph)}
    end
  end

  defp do_topological_sort(graph) do
    vertices = Graph.vertices(graph)

    {result, _} =
      Enum.reduce(vertices, {[], MapSet.new()}, fn vertex, {acc, visited} ->
        if MapSet.member?(visited, vertex) do
          {acc, visited}
        else
          topo_dfs(graph, vertex, visited, acc)
        end
      end)

    result
  end

  defp topo_dfs(graph, vertex, visited, acc) do
    new_visited = MapSet.put(visited, vertex)

    {new_acc, final_visited} =
      graph
      |> Graph.neighbors(vertex)
      |> MapSet.to_list()
      |> Enum.reduce({acc, new_visited}, fn neighbor, {a, v} ->
        if MapSet.member?(v, neighbor) do
          {a, v}
        else
          topo_dfs(graph, neighbor, v, a)
        end
      end)

    {[vertex | new_acc], final_visited}
  end
end

File: haskell/src/Algorithms/DFS.hs
{-|
Module      : Algorithms.DFS
Description : Depth-first search algorithms
Stability   : stable

Depth-first search algorithms for graphs.
Time: O(V + E)
Space: O(V)
-}
module Algorithms.DFS
  ( traverse
  , traverseRecursive
  , findPath
  , hasCycle
  , topologicalSort
  ) where

import qualified Data.Map.Strict as M
import qualified Data.Set as S
import Algorithms.Graph (Graph, neighbors, hasVertex, vertices, isDirected)

-- | DFS traversal from a starting vertex (iterative)
traverse :: Ord a => Graph a -> a -> [a]
traverse g start
  | not (hasVertex start g) = []
  | otherwise = go S.empty [start] []
  where
    go _ [] result = reverse result
    go visited (v:stack) result
      | S.member v visited = go visited stack result
      | otherwise =
          let visited' = S.insert v visited
              newNeighbors = filter (`S.notMember` visited') $ S.toList $ neighbors v g
          in go visited' (newNeighbors ++ stack) (v : result)

-- | DFS traversal from a starting vertex (recursive)
traverseRecursive :: Ord a => Graph a -> a -> [a]
traverseRecursive g start
  | not (hasVertex start g) = []
  | otherwise = go S.empty start
  where
    go visited v
      | S.member v visited = []
      | otherwise =
          let visited' = S.insert v visited
              neighborsList = S.toList $ neighbors v g
          in v : concatMap (go visited') neighborsList

-- | Find a path between two vertices using DFS
findPath :: Ord a => Graph a -> a -> a -> Maybe [a]
findPath g start end
  | not (hasVertex start g) || not (hasVertex end g) = Nothing
  | start == end = Just [start]
  | otherwise = go S.empty [(start, [start])]
  where
    go _ [] = Nothing
    go visited ((v, path):stack)
      | v == end = Just (reverse path)
      | S.member v visited = go visited stack
      | otherwise =
          let visited' = S.insert v visited
              newNeighbors = filter (`S.notMember` visited') $ S.toList $ neighbors v g
              newPaths = [(n, n : path) | n <- newNeighbors]
          in go visited' (newPaths ++ stack)

-- | Check if graph has a cycle
hasCycle :: Ord a => Graph a -> Bool
hasCycle g = any (hasCycleFrom S.empty S.empty) (vertices g)
  where
    hasCycleFrom visited recStack v
      | S.member v recStack = True
      | S.member v visited = False
      | otherwise =
          let visited' = S.insert v visited
              recStack' = S.insert v recStack
              neighborsList = S.toList $ neighbors v g
          in any (hasCycleFrom visited' recStack') neighborsList

-- | Topological sort (only for directed graphs)
topologicalSort :: Ord a => Graph a -> Maybe [a]
topologicalSort g
  | not (isDirected g) = Nothing
  | hasCycle g = Nothing
  | otherwise = Just $ go S.empty [] (vertices g)
  where
    go _ result [] = result
    go visited result (v:vs)
      | S.member v visited = go visited result vs
      | otherwise =
          let (visited', sorted) = dfs visited v
          in go visited' (sorted ++ result) vs

    dfs visited v
      | S.member v visited = (visited, [])
      | otherwise =
          let visited' = S.insert v visited
              neighborsList = S.toList $ neighbors v g
              (visited'', sorted) = foldl dfsNeighbor (visited', []) neighborsList
          in (visited'', v : sorted)

    dfsNeighbor (visited, acc) n =
      let (visited', sorted) = dfs visited n
      in (visited', sorted ++ acc)
